### PR TITLE
deploy: reread hook logs on Retry

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/openshift/origin/issues/9422

@smarterclayton @mfojtik @ironcladlou PTAL

### CODE REVIEW

File: pkg/deploy/strategy/support/lifecycle.go [hunk 1]
```diff
@@ -183,16 +183,40 @@ func (e *HookExecutor) executeExecNewPod(hook *deployapi.LifecycleHook, deployme
...
-			go once.Do(func() { e.readPodLogs(pod, wg) })
+
+			// We should read only the first time or in any container restart when we want to retry.
+			canRetry, restartCount := canRetryReading(updatedPod, restarts)
+			if alreadyRead && !canRetry {
+				break
+			}
+			if restarts != restartCount {
+				wg.Add(1)
+				restarts = restartCount
```
Comment (smarterclayton) [inline pkg/deploy/strategy/support/lifecycle.go]: Would be simpler as `(retry #%d)` and switch on retry 0 vs the others

Comment (smarterclayton) [inline pkg/deploy/strategy/support/lifecycle.go]: Turning things into english is too... clever. :)

**Comment by smarterclayton** [comment]
Minor comment about output, looks good otherwise.

File: pkg/deploy/strategy/support/lifecycle.go (hunk 1 above)
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: > and switch on retry 0 vs the others

Retries map well with the restart count so I think we are covered on switching

**Comment by kargakis** [comment]
Let me get a green run and I will merge it [test]

**Comment by smarterclayton** [comment]
Code doesn't compile.

**Comment by kargakis** [comment]
Extended test is broken. Fixing in a while

**Comment by kargakis** [comment]
extended is green [merge]

**Comment by kargakis** [comment]
Flaked on https://github.com/openshift/origin/issues/9192, [merge]

**Pull request merged**